### DISCUSSION TITLE
Escaping a regex

### DISCUSSION THREAD

**Discussion by strowi** [grafana/tanka#1117] [Q&A] [answered]
Hi,

i have some trouble getting a regex 1:1 from jsonnet to a ConfigMap. Maybe someone here has an idea ;)

This is what i want to end up with inside a file in a pod:

```
topic_pattern: '^abc\.([a-z0-9_-]+)\.([a-z0-9_-]+)\.(data)+$'
...
```

Example:

```
~> tk env add environments/test
~> cat > environments/test/main.jsonnet << EOF
local k = import 'github.com/grafana/jsonnet-libs/ksonnet-util/kausal.libsonnet';
local configMap = k.core.v1.configMap;

{
  configMap:
    configMap.new(
      'config', std.manifestYamlDoc({
        'config.yaml': {
          topic_pattern: '^sl1\\.([a-z0-9_-]+)\\.([a-z0-9_-]+)\\.(data)+$',
        },
      })
    ),
}
EOF

~> tk show environments/test
Error: evaluating jsonnet: /home/rvg/tmp/tt/environments/test/main.jsonnet:9:26-72 Unknown escape sequence in string literal: \.

          topic_pattern: '^sl1\.([a-z0-9_-]+)\.([a-z0-9_-]+)\.(data)+$',
```

No matter what i do with the `\` i always end up with none or duplicate `\`.

Can anyone help me here?

regards,
strowi

**Comment by zerok** [comment]
Are you 100% sure that you don't need the double-backslash? `\.` is not a valid escape inside of a string, so if you want to have the backslash in the content, you need to escape it. 

Weirdly enough, I don't get the same error as you with that example config map and exported it to YAML:

```
apiVersion: v1
data: |-
  "config.yaml":
    "topic_pattern": "^sl1\\.([a-z0-9_-]+)\\.([a-z0-9_-]+)\\.(data)+$"
kind: ConfigMap
metadata:
  name: config
  namespace: disc1117
```

Parsing that will result in only a single backslash:

<img width="631" alt="Screenshot 2024-08-02 at 09 56 18" src="https://github.com/user-attachments/assets/70042212-f9c9-4015-a0c1-cb3ef6ee0e8f">

**Comment by strowi** [comment]
> Are you 100% sure that you don't need the double-backslash? \. is not a valid escape inside of a string, so if you want to have the backslash in the content, you need to escape it.

Thanks, that's what i said ;) We're trying to move stuff from ansible-templates/docker-swarm into Tanka/Kubernetes, where it is a single "\", which i thought was weird. I will insist some more now. ;)

I'll do some more testing with your input.

**Reply by zerok** [reply, marked as answer]
```python
import subprocess
import tempfile
import shutil
import pathlib
import yaml

dir = tempfile.mkdtemp()
try:
    subprocess.check_call('tk init', cwd=dir, shell=True)
    (pathlib.Path(dir) / 'environments' / 'default' / 'main.jsonnet').write_text('''
local k = import 'github.com/grafana/jsonnet-libs/ksonnet-util/kausal.libsonnet';
local configMap = k.core.v1.configMap;

{
  configMap:
    configMap.new(
      'config', std.manifestYamlDoc({
        'config.yaml': {
          topic_pattern: '^sl1\\\\.([a-z0-9_-]+)\\\\.([a-z0-9_-]+)\\\\.(data)+$',
        },
      }),
    ),
}
    ''')
    subprocess.check_call('tk export output environments/default', shell=True, cwd=dir)

    config_map_path = (pathlib.Path(dir) / 'output' / 'v1.ConfigMap-config.yaml')
    with config_map_path.open() as fp:
        config_map = yaml.safe_load(fp)
        print(yaml.safe_load(config_map['data'])['config.yaml']['topic_pattern'])
finally:
    shutil.rmtree(dir)
```

This is a complete and standalone script you can use to replicate it :) Note that in the example there is actually a `\\\\` simply because of wrapping all of that in yet another string 🤣

**Reply by strowi** [reply]
Thx @zerok , it seems "the ansible guy" was wrong after all, though it worked it was missing a "\". ;)
